crypto/kyber/pqcrystals_kyber_ref_common/kem.c

```diff
@@ -1,12 +1,40 @@
 #include "kem.h"
 #include <stddef.h>
 #include <stdint.h>
+#include <string.h>
 #include "indcpa.h"
 #include "params.h"
 #include "symmetric.h"
 #include "verify.h"
 #include "../../rand_extra/pq_custom_randombytes.h"
 
+/*************************************************
+* Name:        crypto_kem_keypair_derand
+*
+* Description: Generates public and private key
+*              for CCA-secure Kyber key encapsulation mechanism
+*
+* Arguments:   - uint8_t *pk: pointer to output public key
+*                (an already allocated array of KYBER_PUBLICKEYBYTES bytes)
+*              - uint8_t *sk: pointer to output private key
+*                (an already allocated array of KYBER_SECRETKEYBYTES bytes)
+*              - uint8_t *coins: pointer to input randomness
+*                (an already allocated array filled with 2*KYBER_SYMBYTES random bytes)
+**
+* Returns 0 (success)
+**************************************************/
+int crypto_kem_keypair_derand(uint8_t *pk,
+                              uint8_t *sk,
+                              const uint8_t *coins)
+{
+  indcpa_keypair_derand(pk, sk, coins);
+  memcpy(sk+KYBER_INDCPA_SECRETKEYBYTES, pk, KYBER_PUBLICKEYBYTES);
+  hash_h(sk+KYBER_SECRETKEYBYTES-2*KYBER_SYMBYTES, pk, KYBER_PUBLICKEYBYTES);
+  /* Value z for pseudo-random output on reject */
+  memcpy(sk+KYBER_SECRETKEYBYTES-KYBER_SYMBYTES, coins+KYBER_SYMBYTES, KYBER_SYMBYTES);
```

Comment:
Also is there a way to test/ensure this is using a separate part of the coins? I verified the firs call up on 30 `indcpa_keypair_derand(pk, sk, coins);` uses KYBER_SYMBYTES and then here you offset that KYBER_SYMBYTES. But this seems fragile and could a future update to indcpa_keypair_derand read more data from the coins and we reuse a part of the coins in a dangerous way? Do you think it's reasonable for `crypto_kem_keypair_derand` to take two coin variables so it's easier to keep them separate? If we messed up this accounting would the KATs break?